{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Loading GEM-PRO dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's try to do a case study with Fumarate...its uniprot ID is P07954"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "What is the uniprot ID?   P07954\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>m_gene_original</th>\n",
       "      <th>m_gene_entrez</th>\n",
       "      <th>m_gene_isoform</th>\n",
       "      <th>u_uniprot_acc</th>\n",
       "      <th>u_isoform_id</th>\n",
       "      <th>u_refseq</th>\n",
       "      <th>u_ensp</th>\n",
       "      <th>u_seq_len</th>\n",
       "      <th>u_seq</th>\n",
       "      <th>u_reviewed</th>\n",
       "      <th>...</th>\n",
       "      <th>ssb_p_aln_coverage</th>\n",
       "      <th>ssb_p_percent_seq_ident</th>\n",
       "      <th>ssb_p_no_deletions_in_pdb</th>\n",
       "      <th>ssb_p_aln_coverage_sim</th>\n",
       "      <th>ssb_si_score</th>\n",
       "      <th>ssb_rez_score</th>\n",
       "      <th>ssb_raw_score</th>\n",
       "      <th>ssb_above_cutoffs</th>\n",
       "      <th>ssb_rank</th>\n",
       "      <th>ssb_best_file</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1223</th>\n",
       "      <td>2271.1</td>\n",
       "      <td>2271</td>\n",
       "      <td>1</td>\n",
       "      <td>P07954</td>\n",
       "      <td>P07954-1</td>\n",
       "      <td>NP_000134</td>\n",
       "      <td>ENSP00000355518</td>\n",
       "      <td>510</td>\n",
       "      <td>MYRALRLLARSRPLVRAPAAALASAPGLGGAAVPSFWPPNAARMAS...</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>459</td>\n",
       "      <td>0.9</td>\n",
       "      <td>True</td>\n",
       "      <td>459</td>\n",
       "      <td>1.44396</td>\n",
       "      <td>1.141667</td>\n",
       "      <td>2.585627</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>3e04.pdb</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     m_gene_original m_gene_entrez m_gene_isoform u_uniprot_acc u_isoform_id  \\\n",
       "1223          2271.1          2271              1        P07954     P07954-1   \n",
       "\n",
       "       u_refseq           u_ensp  u_seq_len  \\\n",
       "1223  NP_000134  ENSP00000355518        510   \n",
       "\n",
       "                                                  u_seq u_reviewed  \\\n",
       "1223  MYRALRLLARSRPLVRAPAAALASAPGLGGAAVPSFWPPNAARMAS...       True   \n",
       "\n",
       "          ...      ssb_p_aln_coverage ssb_p_percent_seq_ident  \\\n",
       "1223      ...                     459                     0.9   \n",
       "\n",
       "     ssb_p_no_deletions_in_pdb ssb_p_aln_coverage_sim ssb_si_score  \\\n",
       "1223                      True                    459      1.44396   \n",
       "\n",
       "     ssb_rez_score ssb_raw_score ssb_above_cutoffs ssb_rank ssb_best_file  \n",
       "1223      1.141667      2.585627              True        1      3e04.pdb  \n",
       "\n",
       "[1 rows x 41 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GP = pd.read_csv('DF_GEMPRO.csv', index_col=0)\n",
    "# forcing gene IDs to be read as strings\n",
    "GP['m_gene_original'] = GP['m_gene_original'].astype(str)\n",
    "GP['m_gene_entrez'] = GP['m_gene_entrez'].astype(str)\n",
    "GP['m_gene_isoform'] = GP['m_gene_isoform'].astype(str)\n",
    "# this searches for an ID and prints out which row it is in\n",
    "gene_id = raw_input(\"What is the uniprot ID?   \") # this can be modified to ask for gene original, entrez, uniprot, isoform id, refseq etc.\n",
    "\n",
    "GP[GP.u_uniprot_acc == gene_id]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Outputs gene of interest, most importantly identifies the best pdb structure. As can be seen, the pdb file is 3e04"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### download the pdb file and get its sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "What is the pdb ID?   3e04\n",
      "Downloading PDB structure '3e04'...\n"
     ]
    }
   ],
   "source": [
    "# load Biopython PDB packages\n",
    "\n",
    "# PDBList to download PDBs\n",
    "from Bio.PDB.PDBList import PDBList\n",
    "pdbl = PDBList()\n",
    "\n",
    "# PDBParser to load and work with files\n",
    "from Bio.PDB.PDBParser import PDBParser\n",
    "parser = PDBParser()\n",
    "\n",
    "import urllib2\n",
    "import uuid\n",
    "\n",
    "pdb_name = raw_input(\"What is the pdb ID?   \")\n",
    "\n",
    "# download pdb\n",
    "pdb_file_path = pdbl.retrieve_pdb_file(pdb_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pdb_structure = parser.get_structure('FUMH', pdb_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['EDO', 'EDO', 'EDO']\n"
     ]
    }
   ],
   "source": [
    "# get the ligands within this file for display\n",
    "# from: http://stackoverflow.com/questions/25718201/remove-heteroatoms-from-pdb\n",
    "ligands = []\n",
    "\n",
    "for residue in pdb_structure.get_residues():\n",
    "    tags = residue.get_full_id()\n",
    "    # tags contains a tuple with (Structure ID, Model ID, Chain ID, (Residue ID))\n",
    "    # Residue ID is a tuple with (*Hetero Field*, Residue ID, Insertion Code)\n",
    "\n",
    "    # Thus you're interested in the Hetero Field, that is empty if the residue\n",
    "    # is not a hetero atom or have some flag if it is (W for waters, H, etc.)\n",
    "    if tags[3][0] != \" \" and tags[3][0] != \"W\":\n",
    "        ligands.append(tags[3][0].split('_')[1].strip())\n",
    "    else:\n",
    "        continue\n",
    "        \n",
    "print(ligands)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Alphabet import IUPAC\n",
    "from Bio.PDB import Polypeptide"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def get_pdb_seq(structure):\n",
    "    '''\n",
    "    Takes in a Biopython structure object and returns a list of the structure's sequences\n",
    "    :param structure: Biopython structure object\n",
    "    :return: Dictionary of sequence strings with chain IDs as the key\n",
    "    '''\n",
    "    \n",
    "    structure_seqs = {}\n",
    "    \n",
    "    # loop over each chain of the PDB\n",
    "    for chain in structure[0]:\n",
    "        \n",
    "        chain_it = iter(chain) \n",
    "        \n",
    "        chain_seq = ''\n",
    "        tracker = 0\n",
    "        \n",
    "        # loop over the residues\n",
    "        for res in chain.get_residues():\n",
    "            # NOTE: you can get the residue number too\n",
    "            res_num = res.id[1]\n",
    "            \n",
    "            # double check if the residue name is a standard residue\n",
    "            # if it is not a standard residue (ie. selenomethionine),\n",
    "            # it will be filled in with an X on the next iteration)\n",
    "            if Polypeptide.is_aa(res, standard=True):\n",
    "                full_id = res.get_full_id()\n",
    "                end_tracker = full_id[3][1]\n",
    "                i_code = full_id[3][2]\n",
    "                aa = Polypeptide.three_to_one(res.get_resname())\n",
    "                \n",
    "                # tracker to fill in X's\n",
    "                if end_tracker != (tracker + 1):# and first == False:\n",
    "                    if i_code != ' ':\n",
    "                        chain_seq += aa\n",
    "                        tracker = end_tracker + 1\n",
    "                        continue\n",
    "                    else:\n",
    "                        chain_seq += 'X'*(end_tracker - tracker - 1)\n",
    "                        \n",
    "                chain_seq += aa\n",
    "                tracker = end_tracker\n",
    "                \n",
    "            else:\n",
    "                continue\n",
    "\n",
    "        structure_seqs[chain.get_id()] = chain_seq\n",
    "\n",
    "    return structure_seqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# represented in a single string\n",
    "pdb_sequence = get_pdb_seq(pdb_structure)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'A': 'XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXSFRIEYDTFGELKVPNDKYYGAQTVRSTMNFKIGGVTERMPTPVIKAFGILKRAAAEVNQDYGLDPKIANAIMKAADEVAEGKLNDHFPLVVWQTGSGTQTNMNVNEVISNRAIEMLGGELGSKIPVHPNDHVNKSQSSNDTFPTAMHIAAAIEVHEVLLPGLQKLHDALDAKSKEFAQIIKIGRTHTQDAVPLTLGQEFSGYVQQVKYAMTRIKAAMPRIYELAAGGTAVGTGLNTRIGFAEKVAAKVAALTGLPFVTAPNKFEALAAHDALVELSGAMNTTACSLMKIANDIRFLGSGPRSGLGELILPENEPGXXXXXGKVNPTQCEAMTMVAAQVMGNHVAVTVGGSNGHFELNVFKPMMIKNVLHSARLLGDASVSFTENCVVGIQANTERINKLMNESLMLVTALNPHIGYDKAAKIAKTAHKNGSTLKETAIELGYLTAEQFDEWVKPKDMLGP',\n",
       " 'B': 'XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXSFRIEYXXXXELKVPNDKYYGAQTVRSTMNFKIGGVTERMPTPVIKAFGILKRAAAEVNQDYGLDPKIANAIMKAADEVAEGKLNDHFPLVVWQTGSGTQTNMNVNEVISNRAIEMLGGELGSKIPVHPNDHVNKSQSSNDTFPTAMHIAAAIEVHEVLLPGLQKLHDALDAKSKEFAQIIKIGRTHTQDAVPLTLGQEFSGYVQQVKYAMTRIKAAMPRIYELAAGGTAVGTGLNTRIGFAEKVAAKVAALTGLPFVTAPNKFEALAAHDALVELSGAMNTTACSLMKIANDIRFLGSGPRSGLGELILPENEPGSSIMPGKVNPTQCEAMTMVAAQVMGNHVAVTVGGSNGHFELNVFKPMMIKNVLHSARLLGDASVSFTENCVVGIQANTERINKLMNESLMLVTALNPHIGXXXXAKIAKTAHKNGSTLKETAIELGYLTAEQFDEWVKPKDMLGPK',\n",
       " 'C': 'XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXSFRIEYDTFGELKVPNDKYYGAQTVRSTMNFKIGGVTERMPTPVIKAFGILKRAAAEVNQDYGLDPKIANAIMKAADEVAEGKLNDHFPLVVWQTGSGTQTNMNVNEVISNRAIEMLGGELGSKIPVHPNDHVNKSQSSNDTFPTAMHIAAAIEVHEVLLPGLQKLHDALDAKSKEFAQIIKIGRTHTQDAVPLTLGQEFSGYVQQVKYAMTRIKAAMPRIYELAAGGTAVGTGLNTRIGFAEKVAAKVAALTGLPFVTAPNKFEALAAHDALVELSGAMNTTACSLMKIANDIRFLGSGPRSGLGELILPENEPGSSIMPGKVNPTQCEAMTMVAAQVMGNHVAVTVGGSNGHFELNVFKPMMIKNVLHSARLLGDASVSFTENCVVGIQANTERINKLMNESLMXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXKPKDMLGPK',\n",
       " 'D': 'XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXSFRIEYDTFGELKVPNDKYYGAQTVRSTMNFKIGGVTERMPTPVIKAFGILKRAAAEVNQDYGLDPKIANAIMKAADEVAEGKLNDHFPLVVWQTGSGTQTNMNVNEVISNRAIEMLGGELGSKIPVHPNDHVNKSQSSNDTFPTAMHIAAAIEVHEVLLPGLQKLHDALDAKSKEFAQIIKIGRTHTQDAVPLTLGQEFSGYVQQVKYAMTRIKAAMPRIYELAAGGTAVGTGLNTRIGFAEKVAAKVAALTGLPFVTAPNKFEALAAHDALVELSGAMNTTACSLMKIANDIRFLGSGPRSGLGELILPENEPGXXXMPGKVNPTQCEAMTMVAAQVMGNHVAVTVGGSNGHFELNVFKPMMIKNVLHSARLLGDASVSFTENCVVGIQANTERINKLMNESLMLVTALNPHIGYDKAAKIAKTAHKNGSTLKETAIELGYLTAEQFDEWVKPKDMLGPK'}"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pdb_sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#temporary\n",
    "fumh_pdb_seq = \"\"\"XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXSFRIEYDTFGELKVPNDKYYGAQTVRSTMNFKIGGVTERMPTPVIKAFGILKRAAAEVNQDYGLDPKIANAIMKAADEVAEGKLNDHFPLVVWQTGSGTQTNMNVNEVISNRAIEMLGGELGSKIPVHPNDHVNKSQSSNDTFPTAMHIAAAIEVHEVLLPGLQKLHDALDAKSKEFAQIIKIGRTHTQDAVPLTLGQEFSGYVQQVKYAMTRIKAAMPRIYELAAGGTAVGTGLNTRIGFAEKVAAKVAALTGLPFVTAPNKFEALAAHDALVELSGAMNTTACSLMKIANDIRFLGSGPRSGLGELILPENEPGXXXXXGKVNPTQCEAMTMVAAQVMGNHVAVTVGGSNGHFELNVFKPMMIKNVLHSARLLGDASVSFTENCVVGIQANTERINKLMNESLMLVTALNPHIGYDKAAKIAKTAHKNGSTLKETAIELGYLTAEQFDEWVKPKDMLGP\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### Writing FASTA files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import os.path\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Alphabet import IUPAC\n",
    "\n",
    "def write_fasta_file(sequence, fileout):\n",
    "    '''\n",
    "    This writes a fasta file for a SeqRecord object. It also checks if the file exists already and returns the filename.\n",
    "    \n",
    "    Input: sequence - Biopython SeqRecord object, identification - ID of the sequence.\n",
    "    Output: Filename of fasta file\n",
    "    '''\n",
    "    \n",
    "    outfile = \"%s\" % fileout\n",
    "    if os.path.isfile(outfile):\n",
    "        print 'FASTA file already exists %s' % outfile\n",
    "        return outfile\n",
    "    else:\n",
    "        SeqIO.write(sequence, outfile, \"fasta\")\n",
    "        return outfile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "What is the gene original ID?   2271.1\n",
      "FASTA file already exists P07954-1.faa\n"
     ]
    }
   ],
   "source": [
    "# example: for gene argss\n",
    "\n",
    "# getting the IDs and making output file name\n",
    "gene_original_id = raw_input('What is the gene original ID?   ') \n",
    "seq_id = GP[GP.m_gene_original == gene_original_id].u_isoform_id.values[0]\n",
    "# the /tmp/ in '/tmp/' + seq_id + '.faa' puts it in a temporary folder; I will remove the temp saving for now\n",
    "seq_output = seq_id + '.faa'\n",
    "\n",
    "# getting the sequence and making it into a Biopython SeqRecord object\n",
    "seq = GP[GP.m_gene_original == gene_original_id].u_seq.values[0]\n",
    "seq_biop = SeqRecord(Seq(seq, IUPAC.protein),id=seq_id,description='uniprot sequence')\n",
    "\n",
    "# writing the SeqRecord object (formats it as FASTA file)\n",
    "out_file = write_fasta_file(seq_biop, seq_output)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">P07954-1 uniprot sequence\r\n",
      "MYRALRLLARSRPLVRAPAAALASAPGLGGAAVPSFWPPNAARMASQNSFRIEYDTFGEL\r\n",
      "KVPNDKYYGAQTVRSTMNFKIGGVTERMPTPVIKAFGILKRAAAEVNQDYGLDPKIANAI\r\n",
      "MKAADEVAEGKLNDHFPLVVWQTGSGTQTNMNVNEVISNRAIEMLGGELGSKIPVHPNDH\r\n",
      "VNKSQSSNDTFPTAMHIAAAIEVHEVLLPGLQKLHDALDAKSKEFAQIIKIGRTHTQDAV\r\n",
      "PLTLGQEFSGYVQQVKYAMTRIKAAMPRIYELAAGGTAVGTGLNTRIGFAEKVAAKVAAL\r\n",
      "TGLPFVTAPNKFEALAAHDALVELSGAMNTTACSLMKIANDIRFLGSGPRSGLGELILPE\r\n",
      "NEPGSSIMPGKVNPTQCEAMTMVAAQVMGNHVAVTVGGSNGHFELNVFKPMMIKNVLHSA\r\n",
      "RLLGDASVSFTENCVVGIQANTERINKLMNESLMLVTALNPHIGYDKAAKIAKTAHKNGS\r\n",
      "TLKETAIELGYLTAEQFDEWVKPKDMLGPK\r\n"
     ]
    }
   ],
   "source": [
    "# just saving in tmp for this example, all fasta files were already written in \"sequence_files\"\n",
    "!cat $out_file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## aligning 2 FASTA files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import os.path\n",
    "from Bio.Emboss.Applications import NeedleCommandline\n",
    "\n",
    "def run_alignment(fasta1_id, fasta1, fasta2_id, fasta2):\n",
    "    '''\n",
    "    Runs the needle alignment program and writes the result to a file. Returns the filename. Standard gap inputs are used.\n",
    "    \n",
    "    Input:  fasta1 - fasta file name (\"reference\" sequence)\n",
    "            fasta2 - fasta file name (what you're interested in aligning)\n",
    "    Output: alignment_file - file name of alignment\n",
    "    '''\n",
    "\n",
    "    alignment_file = \"%s_%s_align.txt\" % (fasta1_id, fasta2_id)\n",
    "    \n",
    "    if os.path.isfile(alignment_file):\n",
    "        print 'Alignment %s file already exists' % alignment_file\n",
    "        return alignment_file\n",
    "\n",
    "    else:\n",
    "        print '**RUNNING ALIGNMENT FOR %s AND %s**' % (fasta1_id, fasta2_id)\n",
    "        needle_cline = NeedleCommandline(asequence=fasta1, bsequence=fasta2, gapopen=10, gapextend=0.5, outfile=alignment_file)\n",
    "        stdout, stderr = needle_cline()\n",
    "        return alignment_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "found uniprot fasta file /Users/LAURENCE/Desktop/Senior Design/Untitled Folder/P07954-1.faa\n",
      "found pdb fasta file /Users/LAURENCE/Desktop/Senior Design/Untitled Folder/3iar.faa\n",
      "**RUNNING ALIGNMENT FOR P07954-1 AND 3iar**\n"
     ]
    }
   ],
   "source": [
    "SEQUENCE_FILES = '/Users/LAURENCE/Desktop/Senior Design/Untitled Folder'\n",
    "UNIPROT_FILES = '/Users/LAURENCE/Desktop/Senior Design/Untitled Folder'\n",
    "PDB_SEQ_FILES = '/Users/LAURENCE/Desktop/Senior Design/Untitled Folder'\n",
    "\n",
    "# 1. get the uniprot sequence file\n",
    "seq_id = GP[GP.m_gene_original == gene_original_id].u_isoform_id.values[0]\n",
    "seq_fasta = os.path.join(UNIPROT_FILES, seq_id + '.faa')\n",
    "\n",
    "if os.path.exists(seq_fasta):\n",
    "    print('found uniprot fasta file {}'.format(seq_fasta))\n",
    "    \n",
    "# 2. get the pdb sequence file\n",
    "pdb_id = GP[GP.m_gene_original == '100.1'].ssb_best_file.values[0].strip('.pdb')\n",
    "pdb_fasta = os.path.join(PDB_SEQ_FILES, pdb_id + '.faa')\n",
    "\n",
    "if os.path.exists(pdb_fasta):\n",
    "    print('found pdb fasta file {}'.format(pdb_fasta))\n",
    "    \n",
    "# 3. run the alignment using the function above\n",
    "os.chdir('/tmp/')\n",
    "alignment_filename = run_alignment(seq_id, seq_fasta, pdb_id, pdb_fasta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "########################################\r\n",
      "# Program: needle\r\n",
      "# Rundate: Thu  4 Feb 2016 17:10:12\r\n",
      "# Commandline: needle\r\n",
      "#    -outfile P07954-1_3iar_align.txt\r\n",
      "#    -asequence \"/Users/LAURENCE/Desktop/Senior Design/Untitled Folder/P07954-1.faa\"\r\n",
      "#    -bsequence \"/Users/LAURENCE/Desktop/Senior Design/Untitled Folder/3iar.faa\"\r\n",
      "#    -gapopen 10\r\n",
      "#    -gapextend 0.5\r\n",
      "# Align_format: srspair\r\n",
      "# Report_file: P07954-1_3iar_align.txt\r\n",
      "########################################\r\n",
      "\r\n",
      "#=======================================\r\n",
      "#\r\n",
      "# Aligned_sequences: 2\r\n",
      "# 1: P07954-1\r\n",
      "# 2: 3iar.A\r\n",
      "# Matrix: EBLOSUM62\r\n",
      "# Gap_penalty: 10.0\r\n",
      "# Extend_penalty: 0.5\r\n",
      "#\r\n",
      "# Length: 623\r\n",
      "# Identity:      65/623 (10.4%)\r\n",
      "# Similarity:   106/623 (17.0%)\r\n",
      "# Gaps:         376/623 (60.4%)\r\n",
      "# Score: 56.0\r\n",
      "# \r\n",
      "#\r\n",
      "#=======================================\r\n",
      "\r\n",
      "P07954-1           1 MYRALRLLARSRPLVRAPAAALASAPGLGGAAVPSFWPPNAARMASQNSF     50\r\n",
      "                                                                       \r\n",
      "3iar.A             0 --------------------------------------------------      0\r\n",
      "\r\n",
      "P07954-1          51 RIEYDTFGELKVPNDKYYGAQTVRSTMNFKIGGVTERMPTPVIKAFGILK    100\r\n",
      "                                                                       \r\n",
      "3iar.A             0 --------------------------------------------------      0\r\n",
      "\r\n",
      "P07954-1         101 RAAAEVNQDYGLDPKIANAIMKAADEVAEGKLNDHFPLVVWQTGSGTQTN    150\r\n",
      "                                                                       \r\n",
      "3iar.A             0 --------------------------------------------------      0\r\n",
      "\r\n",
      "P07954-1         151 MNVNEVISNRAIEMLGGELGSKIPVHPNDHVNKSQSSNDTFPTAMHIAAA    200\r\n",
      "                                                                       \r\n",
      "3iar.A             0 --------------------------------------------------      0\r\n",
      "\r\n",
      "P07954-1         201 IEVHEVLLPGLQK----LHDALDAKSKEFAQIIKIGR-------THTQDA    239\r\n",
      "                             |...|    ||..||...|. ..|:..||       .:|.:.\r\n",
      "3iar.A             1 --------PAFDKPKVELHVHLDGSIKP-ETILYYGRRRGIALPANTAEG     41\r\n",
      "\r\n",
      "P07954-1         240 V--------PLTLGQ---EFSGY----------VQQVKYAMTRIKAAMPR    268\r\n",
      "                     :        ||||..   :|..|          ::::.|....:||....\r\n",
      "3iar.A            42 LLNVIGMDKPLTLPDFLAKFDYYMPAIAGCREAIKRIAYEFVEMKAKEGV     91\r\n",
      "\r\n",
      "P07954-1         269 IY-------ELAAGGTAVGTGLNTRIGFAEKVAAKVAALTGLPFVTAPNK    311\r\n",
      "                     :|       .|.|.........|...|  :....:|.||.|.........\r\n",
      "3iar.A            92 VYVEVRYSPHLLANSKVEPIPWNQAEG--DLTPDEVVALVGQGLQEGERD    139\r\n",
      "\r\n",
      "P07954-1         312 FEALA-------AHD-----ALVELSGAMNTTACSLMKIANDIRFLGSGP    349\r\n",
      "                     |...|       .|.     .:|||...........:.:|.|        \r\n",
      "3iar.A           140 FGVKARSILCCMRHQPNWSPKVVELCKKYQQQTVVAIDLAGD--------    181\r\n",
      "\r\n",
      "P07954-1         350 RSGLGELILPENEPGSSIMPGKVNPTQ----------------------C    377\r\n",
      "                               |..||||::||.|...|                      .\r\n",
      "3iar.A           182 ----------ETIPGSSLLPGHVQAYQEAVKSGIHRTVHAGEVGSAEVVK    221\r\n",
      "\r\n",
      "P07954-1         378 EAMTMVAAQVMGN-------------------------------------    390\r\n",
      "                     ||:.::..:.:|:                                     \r\n",
      "3iar.A           222 EAVDILKTERLGHGYHTLEDQALYNRLRQENMHFEICPWSSYLTGAWKPD    271\r\n",
      "\r\n",
      "P07954-1         391 --HVAVTVGGSNGHFELNVFKPMMIKNVLHSARLLGDASVSFTENCVVGI    438\r\n",
      "                       |..:.:.....::.||...|::.|:.|.:...:....:.|||.....:\r\n",
      "3iar.A           272 TEHAVIRLKNDQANYSLNTDDPLIFKSTLDTDYQMTKRDMGFTEEEFKRL    321\r\n",
      "\r\n",
      "P07954-1         439 QANTERINKL-MNESLMLVTALNPHIGYDKAAKIAKTAHKNGSTLKETAI    487\r\n",
      "                     ..|..:.:.| .:|...|:..|....|...:|...:...           \r\n",
      "3iar.A           322 NINAAKSSFLPEDEKRELLDLLYKAYGMPPSASAGQNLA-----------    360\r\n",
      "\r\n",
      "P07954-1         488 ELGYLTAEQFDEWVKPKDMLGPK    510\r\n",
      "                                            \r\n",
      "3iar.A           360 -----------------------    360\r\n",
      "\r\n",
      "\r\n",
      "#---------------------------------------\r\n",
      "#---------------------------------------\r\n"
     ]
    }
   ],
   "source": [
    "!cat $alignment_filename"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### loading alignment file as a dataframe\n",
    "this code parses a needle alignment file and makes it into a dataframe\n",
    "each row of the dataframe is a position in the reference sequence\n",
    "it also tells you what parts are deleted, mutated, etc."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from Bio import AlignIO\n",
    "from collections import defaultdict\n",
    "\n",
    "def get_alignment_allpos_df(alignment_file, a_seq_id=None, b_seq_id=None):\n",
    "    '''\n",
    "    Takes in a needle alignment file and returns a pandas dataframe of the results\n",
    "    Input: alignment_file - the path to the alignment file, \n",
    "            a_seq_id - optional ID of the reference sequence, \n",
    "            b_seq_id - optional ID of the second sequence\n",
    "    Output: alignment_df - a pandas dataframe of the alignment results\n",
    "    '''\n",
    "    alignments = list(AlignIO.parse(alignment_file, \"emboss\"))\n",
    "\n",
    "    appender = defaultdict(dict)\n",
    "    idx = 0\n",
    "    for alignment in alignments:\n",
    "    #         if not switch:\n",
    "        if not a_seq_id:\n",
    "            a_seq_id = list(alignment)[0].id\n",
    "        a_seq = str(list(alignment)[0].seq)\n",
    "        if not b_seq_id:\n",
    "            b_seq_id = list(alignment)[1].id\n",
    "        b_seq = str(list(alignment)[1].seq)\n",
    "\n",
    "        a_idx = 1\n",
    "        b_idx = 1\n",
    "\n",
    "        for i, (a,b) in enumerate(zip(a_seq,b_seq)):\n",
    "            if a == b and a != '-' and b != '-':\n",
    "                aa_flag = 'match'\n",
    "            if a != b and a == '-' and b != '-':\n",
    "                aa_flag = 'insertion'\n",
    "            if a != b and a != '-' and b == '-':\n",
    "                aa_flag = 'deletion'\n",
    "            if a != b and a != '-' and b == 'X':\n",
    "                aa_flag = 'unresolved'\n",
    "            if a != b and b != '-' and a == 'X':\n",
    "                aa_flag = 'unresolved'\n",
    "            elif a != b and a != '-' and b != '-':\n",
    "                aa_flag = 'mutation'\n",
    "                \n",
    "            appender[idx]['id_a'] = a_seq_id\n",
    "            appender[idx]['id_b'] = b_seq_id\n",
    "            appender[idx]['type'] = aa_flag\n",
    "            \n",
    "            if aa_flag == 'match' or aa_flag == 'unresolved' or aa_flag == 'mutation':\n",
    "                appender[idx]['id_a_aa'] = a\n",
    "                appender[idx]['id_a_pos'] = a_idx\n",
    "                appender[idx]['id_b_aa'] = b\n",
    "                appender[idx]['id_b_pos'] = b_idx\n",
    "                a_idx += 1\n",
    "                b_idx += 1\n",
    "\n",
    "            if aa_flag == 'deletion':\n",
    "                appender[idx]['id_a_aa'] = a\n",
    "                appender[idx]['id_a_pos'] = a_idx\n",
    "                a_idx += 1\n",
    "\n",
    "            if aa_flag == 'insertion':\n",
    "                appender[idx]['id_b_aa'] = b\n",
    "                appender[idx]['id_b_pos'] = b_idx\n",
    "                b_idx += 1\n",
    "            \n",
    "            idx += 1\n",
    "\n",
    "    alignment_df = pd.DataFrame.from_dict(appender, orient='index')\n",
    "    alignment_df = alignment_df[['id_a', 'id_b', 'type', 'id_a_aa', 'id_a_pos', 'id_b_aa', 'id_b_pos']].fillna(value=np.nan)\n",
    "    \n",
    "    return alignment_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id_a</th>\n",
       "      <th>id_b</th>\n",
       "      <th>type</th>\n",
       "      <th>id_a_aa</th>\n",
       "      <th>id_a_pos</th>\n",
       "      <th>id_b_aa</th>\n",
       "      <th>id_b_pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>insertion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>insertion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>insertion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>mutation</td>\n",
       "      <td>M</td>\n",
       "      <td>1</td>\n",
       "      <td>D</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>mutation</td>\n",
       "      <td>S</td>\n",
       "      <td>2</td>\n",
       "      <td>K</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>mutation</td>\n",
       "      <td>S</td>\n",
       "      <td>3</td>\n",
       "      <td>P</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>K</td>\n",
       "      <td>4</td>\n",
       "      <td>K</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>mutation</td>\n",
       "      <td>G</td>\n",
       "      <td>5</td>\n",
       "      <td>V</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>mutation</td>\n",
       "      <td>S</td>\n",
       "      <td>6</td>\n",
       "      <td>E</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>mutation</td>\n",
       "      <td>V</td>\n",
       "      <td>7</td>\n",
       "      <td>L</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>mutation</td>\n",
       "      <td>V</td>\n",
       "      <td>8</td>\n",
       "      <td>H</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>mutation</td>\n",
       "      <td>L</td>\n",
       "      <td>9</td>\n",
       "      <td>V</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>mutation</td>\n",
       "      <td>A</td>\n",
       "      <td>10</td>\n",
       "      <td>H</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>mutation</td>\n",
       "      <td>Y</td>\n",
       "      <td>11</td>\n",
       "      <td>L</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>mutation</td>\n",
       "      <td>S</td>\n",
       "      <td>12</td>\n",
       "      <td>D</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>G</td>\n",
       "      <td>13</td>\n",
       "      <td>G</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>mutation</td>\n",
       "      <td>G</td>\n",
       "      <td>14</td>\n",
       "      <td>S</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>mutation</td>\n",
       "      <td>L</td>\n",
       "      <td>15</td>\n",
       "      <td>I</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>mutation</td>\n",
       "      <td>D</td>\n",
       "      <td>16</td>\n",
       "      <td>K</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>mutation</td>\n",
       "      <td>T</td>\n",
       "      <td>17</td>\n",
       "      <td>P</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>mutation</td>\n",
       "      <td>S</td>\n",
       "      <td>18</td>\n",
       "      <td>E</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>mutation</td>\n",
       "      <td>C</td>\n",
       "      <td>19</td>\n",
       "      <td>T</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>I</td>\n",
       "      <td>20</td>\n",
       "      <td>I</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>L</td>\n",
       "      <td>21</td>\n",
       "      <td>L</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>mutation</td>\n",
       "      <td>V</td>\n",
       "      <td>22</td>\n",
       "      <td>Y</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>mutation</td>\n",
       "      <td>W</td>\n",
       "      <td>23</td>\n",
       "      <td>Y</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>mutation</td>\n",
       "      <td>L</td>\n",
       "      <td>24</td>\n",
       "      <td>G</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>mutation</td>\n",
       "      <td>K</td>\n",
       "      <td>25</td>\n",
       "      <td>R</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>mutation</td>\n",
       "      <td>E</td>\n",
       "      <td>26</td>\n",
       "      <td>R</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>mutation</td>\n",
       "      <td>Q</td>\n",
       "      <td>27</td>\n",
       "      <td>R</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>451</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>insertion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L</td>\n",
       "      <td>331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>452</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>insertion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P</td>\n",
       "      <td>332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>453</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>insertion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>E</td>\n",
       "      <td>333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>454</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>insertion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D</td>\n",
       "      <td>334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>455</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>insertion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>E</td>\n",
       "      <td>335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>456</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>insertion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K</td>\n",
       "      <td>336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>457</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>insertion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>458</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>insertion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>E</td>\n",
       "      <td>338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>459</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>insertion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L</td>\n",
       "      <td>339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>460</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>insertion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L</td>\n",
       "      <td>340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>insertion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D</td>\n",
       "      <td>341</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>462</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>insertion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L</td>\n",
       "      <td>342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>insertion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L</td>\n",
       "      <td>343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>464</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>insertion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Y</td>\n",
       "      <td>344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>465</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>insertion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K</td>\n",
       "      <td>345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>466</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>insertion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>467</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>insertion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Y</td>\n",
       "      <td>347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>468</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>insertion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>469</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>insertion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>M</td>\n",
       "      <td>349</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>470</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>insertion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P</td>\n",
       "      <td>350</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>471</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>insertion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P</td>\n",
       "      <td>351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>472</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>insertion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S</td>\n",
       "      <td>352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>473</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>insertion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>474</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>insertion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S</td>\n",
       "      <td>354</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>475</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>insertion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>476</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>insertion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>477</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>insertion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Q</td>\n",
       "      <td>357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>478</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>insertion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>N</td>\n",
       "      <td>358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>479</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>insertion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L</td>\n",
       "      <td>359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>480</th>\n",
       "      <td>P00966-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>insertion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>360</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>481 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         id_a    id_b       type id_a_aa  id_a_pos id_b_aa  id_b_pos\n",
       "0    P00966-1  3iar.A  insertion     NaN       NaN       P         1\n",
       "1    P00966-1  3iar.A  insertion     NaN       NaN       A         2\n",
       "2    P00966-1  3iar.A  insertion     NaN       NaN       F         3\n",
       "3    P00966-1  3iar.A   mutation       M         1       D         4\n",
       "4    P00966-1  3iar.A   mutation       S         2       K         5\n",
       "5    P00966-1  3iar.A   mutation       S         3       P         6\n",
       "6    P00966-1  3iar.A      match       K         4       K         7\n",
       "7    P00966-1  3iar.A   mutation       G         5       V         8\n",
       "8    P00966-1  3iar.A   mutation       S         6       E         9\n",
       "9    P00966-1  3iar.A   mutation       V         7       L        10\n",
       "10   P00966-1  3iar.A   mutation       V         8       H        11\n",
       "11   P00966-1  3iar.A   mutation       L         9       V        12\n",
       "12   P00966-1  3iar.A   mutation       A        10       H        13\n",
       "13   P00966-1  3iar.A   mutation       Y        11       L        14\n",
       "14   P00966-1  3iar.A   mutation       S        12       D        15\n",
       "15   P00966-1  3iar.A      match       G        13       G        16\n",
       "16   P00966-1  3iar.A   mutation       G        14       S        17\n",
       "17   P00966-1  3iar.A   mutation       L        15       I        18\n",
       "18   P00966-1  3iar.A   mutation       D        16       K        19\n",
       "19   P00966-1  3iar.A   mutation       T        17       P        20\n",
       "20   P00966-1  3iar.A   mutation       S        18       E        21\n",
       "21   P00966-1  3iar.A   mutation       C        19       T        22\n",
       "22   P00966-1  3iar.A      match       I        20       I        23\n",
       "23   P00966-1  3iar.A      match       L        21       L        24\n",
       "24   P00966-1  3iar.A   mutation       V        22       Y        25\n",
       "25   P00966-1  3iar.A   mutation       W        23       Y        26\n",
       "26   P00966-1  3iar.A   mutation       L        24       G        27\n",
       "27   P00966-1  3iar.A   mutation       K        25       R        28\n",
       "28   P00966-1  3iar.A   mutation       E        26       R        29\n",
       "29   P00966-1  3iar.A   mutation       Q        27       R        30\n",
       "..        ...     ...        ...     ...       ...     ...       ...\n",
       "451  P00966-1  3iar.A  insertion     NaN       NaN       L       331\n",
       "452  P00966-1  3iar.A  insertion     NaN       NaN       P       332\n",
       "453  P00966-1  3iar.A  insertion     NaN       NaN       E       333\n",
       "454  P00966-1  3iar.A  insertion     NaN       NaN       D       334\n",
       "455  P00966-1  3iar.A  insertion     NaN       NaN       E       335\n",
       "456  P00966-1  3iar.A  insertion     NaN       NaN       K       336\n",
       "457  P00966-1  3iar.A  insertion     NaN       NaN       R       337\n",
       "458  P00966-1  3iar.A  insertion     NaN       NaN       E       338\n",
       "459  P00966-1  3iar.A  insertion     NaN       NaN       L       339\n",
       "460  P00966-1  3iar.A  insertion     NaN       NaN       L       340\n",
       "461  P00966-1  3iar.A  insertion     NaN       NaN       D       341\n",
       "462  P00966-1  3iar.A  insertion     NaN       NaN       L       342\n",
       "463  P00966-1  3iar.A  insertion     NaN       NaN       L       343\n",
       "464  P00966-1  3iar.A  insertion     NaN       NaN       Y       344\n",
       "465  P00966-1  3iar.A  insertion     NaN       NaN       K       345\n",
       "466  P00966-1  3iar.A  insertion     NaN       NaN       A       346\n",
       "467  P00966-1  3iar.A  insertion     NaN       NaN       Y       347\n",
       "468  P00966-1  3iar.A  insertion     NaN       NaN       G       348\n",
       "469  P00966-1  3iar.A  insertion     NaN       NaN       M       349\n",
       "470  P00966-1  3iar.A  insertion     NaN       NaN       P       350\n",
       "471  P00966-1  3iar.A  insertion     NaN       NaN       P       351\n",
       "472  P00966-1  3iar.A  insertion     NaN       NaN       S       352\n",
       "473  P00966-1  3iar.A  insertion     NaN       NaN       A       353\n",
       "474  P00966-1  3iar.A  insertion     NaN       NaN       S       354\n",
       "475  P00966-1  3iar.A  insertion     NaN       NaN       A       355\n",
       "476  P00966-1  3iar.A  insertion     NaN       NaN       G       356\n",
       "477  P00966-1  3iar.A  insertion     NaN       NaN       Q       357\n",
       "478  P00966-1  3iar.A  insertion     NaN       NaN       N       358\n",
       "479  P00966-1  3iar.A  insertion     NaN       NaN       L       359\n",
       "480  P00966-1  3iar.A  insertion     NaN       NaN       A       360\n",
       "\n",
       "[481 rows x 7 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_alignment_allpos_df(alignment_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
