{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Loading GEM-PRO dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>m_gene_original</th>\n",
       "      <th>m_gene_entrez</th>\n",
       "      <th>m_gene_isoform</th>\n",
       "      <th>u_uniprot_acc</th>\n",
       "      <th>u_isoform_id</th>\n",
       "      <th>u_refseq</th>\n",
       "      <th>u_ensp</th>\n",
       "      <th>u_seq_len</th>\n",
       "      <th>u_seq</th>\n",
       "      <th>u_reviewed</th>\n",
       "      <th>...</th>\n",
       "      <th>ssb_p_aln_coverage</th>\n",
       "      <th>ssb_p_percent_seq_ident</th>\n",
       "      <th>ssb_p_no_deletions_in_pdb</th>\n",
       "      <th>ssb_p_aln_coverage_sim</th>\n",
       "      <th>ssb_si_score</th>\n",
       "      <th>ssb_rez_score</th>\n",
       "      <th>ssb_raw_score</th>\n",
       "      <th>ssb_above_cutoffs</th>\n",
       "      <th>ssb_rank</th>\n",
       "      <th>ssb_best_file</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>100.1</td>\n",
       "      <td>100</td>\n",
       "      <td>1</td>\n",
       "      <td>P00813</td>\n",
       "      <td>P00813-1</td>\n",
       "      <td>NP_000013</td>\n",
       "      <td>ENSP00000361965</td>\n",
       "      <td>363</td>\n",
       "      <td>MAQTPAFDKPKVELHVHLDGSIKPETILYYGRRRGIALPANTAEGL...</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>359</td>\n",
       "      <td>0.988981</td>\n",
       "      <td>True</td>\n",
       "      <td>359</td>\n",
       "      <td>1.586721</td>\n",
       "      <td>1.213333</td>\n",
       "      <td>2.800054</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>3iar.pdb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10005.1</td>\n",
       "      <td>10005</td>\n",
       "      <td>1</td>\n",
       "      <td>O14734</td>\n",
       "      <td>O14734-1</td>\n",
       "      <td>NP_005460</td>\n",
       "      <td>ENSP00000217455</td>\n",
       "      <td>319</td>\n",
       "      <td>MSSPQAPEDGQGCGDRGDPPGDLRSVLVTTVLNLEPLDEDLFRGRH...</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NP_005460.2_model1_fix.pdb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10005.2</td>\n",
       "      <td>10005</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10005.3</td>\n",
       "      <td>10005</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10007.1</td>\n",
       "      <td>10007</td>\n",
       "      <td>1</td>\n",
       "      <td>P46926</td>\n",
       "      <td>P46926-1</td>\n",
       "      <td>NP_005462</td>\n",
       "      <td>ENSP00000311876</td>\n",
       "      <td>289</td>\n",
       "      <td>MKLIILEHYSQASEWAAKYIRNRIIQFNPGPEKYFTLGLPTGSTPL...</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>281</td>\n",
       "      <td>0.972318</td>\n",
       "      <td>True</td>\n",
       "      <td>281</td>\n",
       "      <td>1.559988</td>\n",
       "      <td>1.175000</td>\n",
       "      <td>2.734988</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>1ne7.pdb</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  m_gene_original m_gene_entrez m_gene_isoform u_uniprot_acc u_isoform_id  \\\n",
       "0           100.1           100              1        P00813     P00813-1   \n",
       "1         10005.1         10005              1        O14734     O14734-1   \n",
       "2         10005.2         10005              2           NaN          NaN   \n",
       "3         10005.3         10005              3           NaN          NaN   \n",
       "4         10007.1         10007              1        P46926     P46926-1   \n",
       "\n",
       "    u_refseq           u_ensp  u_seq_len  \\\n",
       "0  NP_000013  ENSP00000361965        363   \n",
       "1  NP_005460  ENSP00000217455        319   \n",
       "2        NaN              NaN        NaN   \n",
       "3        NaN              NaN        NaN   \n",
       "4  NP_005462  ENSP00000311876        289   \n",
       "\n",
       "                                               u_seq u_reviewed  \\\n",
       "0  MAQTPAFDKPKVELHVHLDGSIKPETILYYGRRRGIALPANTAEGL...       True   \n",
       "1  MSSPQAPEDGQGCGDRGDPPGDLRSVLVTTVLNLEPLDEDLFRGRH...       True   \n",
       "2                                                NaN        NaN   \n",
       "3                                                NaN        NaN   \n",
       "4  MKLIILEHYSQASEWAAKYIRNRIIQFNPGPEKYFTLGLPTGSTPL...       True   \n",
       "\n",
       "              ...             ssb_p_aln_coverage ssb_p_percent_seq_ident  \\\n",
       "0             ...                            359                0.988981   \n",
       "1             ...                            NaN                     NaN   \n",
       "2             ...                            NaN                     NaN   \n",
       "3             ...                            NaN                     NaN   \n",
       "4             ...                            281                0.972318   \n",
       "\n",
       "  ssb_p_no_deletions_in_pdb ssb_p_aln_coverage_sim ssb_si_score ssb_rez_score  \\\n",
       "0                      True                    359     1.586721      1.213333   \n",
       "1                       NaN                    NaN          NaN           NaN   \n",
       "2                       NaN                    NaN          NaN           NaN   \n",
       "3                       NaN                    NaN          NaN           NaN   \n",
       "4                      True                    281     1.559988      1.175000   \n",
       "\n",
       "  ssb_raw_score ssb_above_cutoffs ssb_rank               ssb_best_file  \n",
       "0      2.800054              True        1                    3iar.pdb  \n",
       "1           NaN               NaN      NaN  NP_005460.2_model1_fix.pdb  \n",
       "2           NaN               NaN      NaN                         NaN  \n",
       "3           NaN               NaN      NaN                         NaN  \n",
       "4      2.734988              True        1                    1ne7.pdb  \n",
       "\n",
       "[5 rows x 41 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GP = pd.read_csv('DF_GEMPRO.csv', index_col=0)\n",
    "# forcing gene IDs to be read as strings\n",
    "GP['m_gene_original'] = GP['m_gene_original'].astype(str)\n",
    "GP['m_gene_entrez'] = GP['m_gene_entrez'].astype(str)\n",
    "GP['m_gene_isoform'] = GP['m_gene_isoform'].astype(str)\n",
    "GP['m_gene_original'][1]\n",
    "GP.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Note the excel file has the sequence IDs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Outputs gene of interest, most importantly identifies the best pdb structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "What is the gene ID?   100\n",
      "                                                                           0\n",
      "m_gene_original                                                        100.1\n",
      "m_gene_entrez                                                            100\n",
      "m_gene_isoform                                                             1\n",
      "u_uniprot_acc                                                         P00813\n",
      "u_isoform_id                                                        P00813-1\n",
      "u_refseq                                                           NP_000013\n",
      "u_ensp                                                       ENSP00000361965\n",
      "u_seq_len                                                                363\n",
      "u_seq                      MAQTPAFDKPKVELHVHLDGSIKPETILYYGRRRGIALPANTAEGL...\n",
      "u_reviewed                                                              True\n",
      "u_gene_name                                                              ADA\n",
      "u_ec_number                                                      ['3.5.4.4']\n",
      "u_description              ['Adenosine aminohydrolase', 'Adenosine deamin...\n",
      "u_entry_version                                                   2016-01-20\n",
      "u_seq_version                                                     2007-01-23\n",
      "u_pfam                                                           ['PF00962']\n",
      "u_kegg_id                                                        ['hsa:100']\n",
      "u_go                       ['GO:0030054; C:cell junction; IEA:UniProtKB-S...\n",
      "i_entry_name                                      NP_000013.2_model1_fix.pdb\n",
      "u_pdb                                                                   3IAR\n",
      "p_experiment                                               X-RAY DIFFRACTION\n",
      "p_resolution                                                            1.52\n",
      "p_resolution_scaled                                                 0.894349\n",
      "p_chemicals                                      ['NI', 'GOL', 'NO3', '3D1']\n",
      "p_chains                                                               ['A']\n",
      "p_chain_uniprot_map                                        {'A': ['P00813']}\n",
      "p_ec_numbers                                            [('A', ['3.5.4.4'])]\n",
      "p_deposition_date                                                  14-JUL-09\n",
      "p_doi                                                                    NaN\n",
      "p_pmid                                                                   NaN\n",
      "p_space_group                                                     P 21 21 21\n",
      "ssb_p_aln_coverage                                                       359\n",
      "ssb_p_percent_seq_ident                                             0.988981\n",
      "ssb_p_no_deletions_in_pdb                                               True\n",
      "ssb_p_aln_coverage_sim                                                   359\n",
      "ssb_si_score                                                         1.58672\n",
      "ssb_rez_score                                                        1.21333\n",
      "ssb_raw_score                                                        2.80005\n",
      "ssb_above_cutoffs                                                       True\n",
      "ssb_rank                                                                   1\n",
      "ssb_best_file                                                       3iar.pdb\n"
     ]
    }
   ],
   "source": [
    "# this searches for an ID and prints out which row it is in\n",
    "gene_id = raw_input(\"What is the gene ID?   \") # this can be modified to ask for gene original, entrez, uniprot, isoform id, refseq etc.\n",
    "\n",
    "index = 0 \n",
    "for ID in  GP['m_gene_entrez']:\n",
    "    if ID == gene_id:\n",
    "        print pd.DataFrame(GP.ix[index])\n",
    "    index += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### Writing FASTA files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import os.path\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Alphabet import IUPAC\n",
    "\n",
    "def write_fasta_file(sequence, fileout):\n",
    "    '''\n",
    "    This writes a fasta file for a SeqRecord object. It also checks if the file exists already and returns the filename.\n",
    "    \n",
    "    Input: sequence - Biopython SeqRecord object, identification - ID of the sequence.\n",
    "    Output: Filename of fasta file\n",
    "    '''\n",
    "    \n",
    "    outfile = \"%s\" % fileout\n",
    "    if os.path.isfile(outfile):\n",
    "        print 'FASTA file already exists %s' % outfile\n",
    "        return outfile\n",
    "    else:\n",
    "        SeqIO.write(sequence, outfile, \"fasta\")\n",
    "        return outfile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_pdb_seq(structure):\n",
    "    '''\n",
    "    Takes in a Biopython structure object and returns a list of the structure's sequences\n",
    "    :param structure: Biopython structure object\n",
    "    :return: Dictionary of sequence strings with chain IDs as the key\n",
    "    '''\n",
    "    \n",
    "    structure_seqs = {}\n",
    "    \n",
    "    # loop over each chain of the PDB\n",
    "    for chain in structure[0]:\n",
    "        \n",
    "        chain_it = iter(chain) \n",
    "        \n",
    "        chain_seq = ''\n",
    "        tracker = 0\n",
    "        \n",
    "        # loop over the residues\n",
    "        for res in chain.get_residues():\n",
    "            # NOTE: you can get the residue number too\n",
    "            res_num = res.id[1]\n",
    "            \n",
    "            # double check if the residue name is a standard residue\n",
    "            # if it is not a standard residue (ie. selenomethionine),\n",
    "            # it will be filled in with an X on the next iteration)\n",
    "            if Polypeptide.is_aa(res, standard=True):\n",
    "                full_id = res.get_full_id()\n",
    "                end_tracker = full_id[3][1]\n",
    "                i_code = full_id[3][2]\n",
    "                aa = Polypeptide.three_to_one(res.get_resname())\n",
    "                \n",
    "                # tracker to fill in X's\n",
    "                if end_tracker != (tracker + 1):# and first == False:\n",
    "                    if i_code != ' ':\n",
    "                        chain_seq += aa\n",
    "                        tracker = end_tracker + 1\n",
    "                        continue\n",
    "                    else:\n",
    "                        chain_seq += 'X'*(end_tracker - tracker - 1)\n",
    "                        \n",
    "                chain_seq += aa\n",
    "                tracker = end_tracker\n",
    "                \n",
    "            else:\n",
    "                continue\n",
    "\n",
    "        structure_seqs[chain.get_id()] = chain_seq\n",
    "\n",
    "    return structure_seqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FASTA file already exists P00813-1.faa\n"
     ]
    }
   ],
   "source": [
    "# example: for gene 100.1\n",
    "\n",
    "# getting the IDs and making output file name\n",
    "seq_id = GP[GP.m_gene_original == '100.1'].u_isoform_id.values[0]\n",
    "# the /tmp/ in '/tmp/' + seq_id + '.faa' puts it in a temporary folder; I will remove the temp saving for now\n",
    "seq_output = seq_id + '.faa'\n",
    "\n",
    "# getting the sequence and making it into a Biopython SeqRecord object\n",
    "seq = GP[GP.m_gene_original == '100.1'].u_seq.values[0]\n",
    "seq_biop = SeqRecord(Seq(seq, IUPAC.protein),id=seq_id,description='uniprot sequence')\n",
    "\n",
    "# writing the SeqRecord object (formats it as FASTA file)\n",
    "out_file = write_fasta_file(seq_biop, seq_output)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">P00813-1 uniprot sequence\r\n",
      "MAQTPAFDKPKVELHVHLDGSIKPETILYYGRRRGIALPANTAEGLLNVIGMDKPLTLPD\r\n",
      "FLAKFDYYMPAIAGCREAIKRIAYEFVEMKAKEGVVYVEVRYSPHLLANSKVEPIPWNQA\r\n",
      "EGDLTPDEVVALVGQGLQEGERDFGVKARSILCCMRHQPNWSPKVVELCKKYQQQTVVAI\r\n",
      "DLAGDETIPGSSLLPGHVQAYQEAVKSGIHRTVHAGEVGSAEVVKEAVDILKTERLGHGY\r\n",
      "HTLEDQALYNRLRQENMHFEICPWSSYLTGAWKPDTEHAVIRLKNDQANYSLNTDDPLIF\r\n",
      "KSTLDTDYQMTKRDMGFTEEEFKRLNINAAKSSFLPEDEKRELLDLLYKAYGMPPSASAG\r\n",
      "QNL\r\n"
     ]
    }
   ],
   "source": [
    "# just saving in tmp for this example, all fasta files were already written in \"sequence_files\"\n",
    "!cat $out_file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## aligning 2 FASTA files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import os.path\n",
    "from Bio.Emboss.Applications import NeedleCommandline\n",
    "\n",
    "def run_alignment(fasta1_id, fasta1, fasta2_id, fasta2):\n",
    "    '''\n",
    "    Runs the needle alignment program and writes the result to a file. Returns the filename. Standard gap inputs are used.\n",
    "    \n",
    "    Input:  fasta1 - fasta file name (\"reference\" sequence)\n",
    "            fasta2 - fasta file name (what you're interested in aligning)\n",
    "    Output: alignment_file - file name of alignment\n",
    "    '''\n",
    "\n",
    "    alignment_file = \"%s_%s_align.txt\" % (fasta1_id, fasta2_id)\n",
    "    \n",
    "    if os.path.isfile(alignment_file):\n",
    "        print 'Alignment %s file already exists' % alignment_file\n",
    "        return alignment_file\n",
    "\n",
    "    else:\n",
    "        print '**RUNNING ALIGNMENT FOR %s AND %s**' % (fasta1_id, fasta2_id)\n",
    "        needle_cline = NeedleCommandline(asequence=fasta1, bsequence=fasta2, gapopen=10, gapextend=0.5, outfile=alignment_file)\n",
    "        stdout, stderr = needle_cline()\n",
    "        return alignment_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Structure exists: '/Users/LAURENCE/Desktop/Senior Design/Untitled Folder/ia/pdb3iar.ent' \n"
     ]
    }
   ],
   "source": [
    "# load Biopython PDB packages\n",
    "\n",
    "# PDBList to download PDBs\n",
    "from Bio.PDB.PDBList import PDBList\n",
    "pdbl = PDBList()\n",
    "\n",
    "# PDBParser to load and work with files\n",
    "from Bio.PDB.PDBParser import PDBParser\n",
    "parser = PDBParser()\n",
    "\n",
    "import urllib2\n",
    "import uuid\n",
    "\n",
    "# download pdb\n",
    "pdb_file_path = pdbl.retrieve_pdb_file('3iar')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from Bio.Emboss.Applications import NeedleCommandline\n",
    "# I manually downloaded 3iar.faa from Nathan's dropbox\n",
    "needle_cline = NeedleCommandline(asequence=\"3iar.faa\", bsequence=\"P00813-1.faa\",\n",
    "                                gapopen=10, gapextend=0.5, outfile=\"needle.txt\")\n",
    "stdout,stderr=needle_cline()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SingleLetterAlphabet() alignment with 2 rows and 364 columns\n",
      "----PAFDKPKVELHVHLDGSIKPETILYYGRRRGIALPANTAE...NLA 3iar.A\n",
      "MAQTPAFDKPKVELHVHLDGSIKPETILYYGRRRGIALPANTAE...NL- P00813-1\n"
     ]
    }
   ],
   "source": [
    "from Bio import AlignIO\n",
    "align = AlignIO.read(\"needle.txt\",\"emboss\")\n",
    "print (align)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "found uniprot fasta file /Users/LAURENCE/Desktop/Senior Design/Untitled Folder/P00813-1.faa\n",
      "found pdb fasta file /Users/LAURENCE/Desktop/Senior Design/Untitled Folder/3iar.faa\n",
      "**RUNNING ALIGNMENT FOR P00813-1 AND 3iar**\n"
     ]
    }
   ],
   "source": [
    "# example: aligning 100.1 to 3iar.pdb\n",
    "\n",
    "SEQUENCE_FILES = '/Users/LAURENCE/Desktop/Senior Design/Untitled Folder'\n",
    "UNIPROT_FILES = '/Users/LAURENCE/Desktop/Senior Design/Untitled Folder'\n",
    "PDB_SEQ_FILES = '/Users/LAURENCE/Desktop/Senior Design/Untitled Folder'\n",
    "\n",
    "# 1. get the uniprot sequence file\n",
    "seq_id = GP[GP.m_gene_original == '100.1'].u_isoform_id.values[0]\n",
    "seq_fasta = os.path.join(UNIPROT_FILES, seq_id + '.faa')\n",
    "\n",
    "if os.path.exists(seq_fasta):\n",
    "    print('found uniprot fasta file {}'.format(seq_fasta))\n",
    "    \n",
    "# 2. get the pdb sequence file\n",
    "pdb_id = GP[GP.m_gene_original == '100.1'].ssb_best_file.values[0].strip('.pdb')\n",
    "pdb_fasta = os.path.join(PDB_SEQ_FILES, pdb_id + '.faa')\n",
    "\n",
    "if os.path.exists(pdb_fasta):\n",
    "    print('found pdb fasta file {}'.format(pdb_fasta))\n",
    "    \n",
    "# 3. run the alignment using the function above\n",
    "os.chdir('/tmp/')\n",
    "alignment_filename = run_alignment(seq_id, seq_fasta, pdb_id, pdb_fasta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "########################################\r\n",
      "# Program: needle\r\n",
      "# Rundate: Thu  4 Feb 2016 14:15:56\r\n",
      "# Commandline: needle\r\n",
      "#    -outfile P00813-1_3iar_align.txt\r\n",
      "#    -asequence \"/Users/LAURENCE/Desktop/Senior Design/Untitled Folder/P00813-1.faa\"\r\n",
      "#    -bsequence \"/Users/LAURENCE/Desktop/Senior Design/Untitled Folder/3iar.faa\"\r\n",
      "#    -gapopen 10\r\n",
      "#    -gapextend 0.5\r\n",
      "# Align_format: srspair\r\n",
      "# Report_file: P00813-1_3iar_align.txt\r\n",
      "########################################\r\n",
      "\r\n",
      "#=======================================\r\n",
      "#\r\n",
      "# Aligned_sequences: 2\r\n",
      "# 1: P00813-1\r\n",
      "# 2: 3iar.A\r\n",
      "# Matrix: EBLOSUM62\r\n",
      "# Gap_penalty: 10.0\r\n",
      "# Extend_penalty: 0.5\r\n",
      "#\r\n",
      "# Length: 364\r\n",
      "# Identity:     359/364 (98.6%)\r\n",
      "# Similarity:   359/364 (98.6%)\r\n",
      "# Gaps:           5/364 ( 1.4%)\r\n",
      "# Score: 1889.0\r\n",
      "# \r\n",
      "#\r\n",
      "#=======================================\r\n",
      "\r\n",
      "P00813-1           1 MAQTPAFDKPKVELHVHLDGSIKPETILYYGRRRGIALPANTAEGLLNVI     50\r\n",
      "                         ||||||||||||||||||||||||||||||||||||||||||||||\r\n",
      "3iar.A             1 ----PAFDKPKVELHVHLDGSIKPETILYYGRRRGIALPANTAEGLLNVI     46\r\n",
      "\r\n",
      "P00813-1          51 GMDKPLTLPDFLAKFDYYMPAIAGCREAIKRIAYEFVEMKAKEGVVYVEV    100\r\n",
      "                     ||||||||||||||||||||||||||||||||||||||||||||||||||\r\n",
      "3iar.A            47 GMDKPLTLPDFLAKFDYYMPAIAGCREAIKRIAYEFVEMKAKEGVVYVEV     96\r\n",
      "\r\n",
      "P00813-1         101 RYSPHLLANSKVEPIPWNQAEGDLTPDEVVALVGQGLQEGERDFGVKARS    150\r\n",
      "                     ||||||||||||||||||||||||||||||||||||||||||||||||||\r\n",
      "3iar.A            97 RYSPHLLANSKVEPIPWNQAEGDLTPDEVVALVGQGLQEGERDFGVKARS    146\r\n",
      "\r\n",
      "P00813-1         151 ILCCMRHQPNWSPKVVELCKKYQQQTVVAIDLAGDETIPGSSLLPGHVQA    200\r\n",
      "                     ||||||||||||||||||||||||||||||||||||||||||||||||||\r\n",
      "3iar.A           147 ILCCMRHQPNWSPKVVELCKKYQQQTVVAIDLAGDETIPGSSLLPGHVQA    196\r\n",
      "\r\n",
      "P00813-1         201 YQEAVKSGIHRTVHAGEVGSAEVVKEAVDILKTERLGHGYHTLEDQALYN    250\r\n",
      "                     ||||||||||||||||||||||||||||||||||||||||||||||||||\r\n",
      "3iar.A           197 YQEAVKSGIHRTVHAGEVGSAEVVKEAVDILKTERLGHGYHTLEDQALYN    246\r\n",
      "\r\n",
      "P00813-1         251 RLRQENMHFEICPWSSYLTGAWKPDTEHAVIRLKNDQANYSLNTDDPLIF    300\r\n",
      "                     ||||||||||||||||||||||||||||||||||||||||||||||||||\r\n",
      "3iar.A           247 RLRQENMHFEICPWSSYLTGAWKPDTEHAVIRLKNDQANYSLNTDDPLIF    296\r\n",
      "\r\n",
      "P00813-1         301 KSTLDTDYQMTKRDMGFTEEEFKRLNINAAKSSFLPEDEKRELLDLLYKA    350\r\n",
      "                     ||||||||||||||||||||||||||||||||||||||||||||||||||\r\n",
      "3iar.A           297 KSTLDTDYQMTKRDMGFTEEEFKRLNINAAKSSFLPEDEKRELLDLLYKA    346\r\n",
      "\r\n",
      "P00813-1         351 YGMPPSASAGQNL-    363\r\n",
      "                     ||||||||||||| \r\n",
      "3iar.A           347 YGMPPSASAGQNLA    360\r\n",
      "\r\n",
      "\r\n",
      "#---------------------------------------\r\n",
      "#---------------------------------------\r\n"
     ]
    }
   ],
   "source": [
    "!cat $alignment_filename"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### loading alignment file as a dataframe\n",
    "this code parses a needle alignment file and makes it into a dataframe\n",
    "each row of the dataframe is a position in the reference sequence\n",
    "it also tells you what parts are deleted, mutated, etc."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from Bio import AlignIO\n",
    "from collections import defaultdict\n",
    "\n",
    "def get_alignment_allpos_df(alignment_file, a_seq_id=None, b_seq_id=None):\n",
    "    '''\n",
    "    Takes in a needle alignment file and returns a pandas dataframe of the results\n",
    "    Input: alignment_file - the path to the alignment file, \n",
    "            a_seq_id - optional ID of the reference sequence, \n",
    "            b_seq_id - optional ID of the second sequence\n",
    "    Output: alignment_df - a pandas dataframe of the alignment results\n",
    "    '''\n",
    "    alignments = list(AlignIO.parse(alignment_file, \"emboss\"))\n",
    "\n",
    "    appender = defaultdict(dict)\n",
    "    idx = 0\n",
    "    for alignment in alignments:\n",
    "    #         if not switch:\n",
    "        if not a_seq_id:\n",
    "            a_seq_id = list(alignment)[0].id\n",
    "        a_seq = str(list(alignment)[0].seq)\n",
    "        if not b_seq_id:\n",
    "            b_seq_id = list(alignment)[1].id\n",
    "        b_seq = str(list(alignment)[1].seq)\n",
    "\n",
    "        a_idx = 1\n",
    "        b_idx = 1\n",
    "\n",
    "        for i, (a,b) in enumerate(zip(a_seq,b_seq)):\n",
    "            if a == b and a != '-' and b != '-':\n",
    "                aa_flag = 'match'\n",
    "            if a != b and a == '-' and b != '-':\n",
    "                aa_flag = 'insertion'\n",
    "            if a != b and a != '-' and b == '-':\n",
    "                aa_flag = 'deletion'\n",
    "            if a != b and a != '-' and b == 'X':\n",
    "                aa_flag = 'unresolved'\n",
    "            if a != b and b != '-' and a == 'X':\n",
    "                aa_flag = 'unresolved'\n",
    "            elif a != b and a != '-' and b != '-':\n",
    "                aa_flag = 'mutation'\n",
    "                \n",
    "            appender[idx]['id_a'] = a_seq_id\n",
    "            appender[idx]['id_b'] = b_seq_id\n",
    "            appender[idx]['type'] = aa_flag\n",
    "            \n",
    "            if aa_flag == 'match' or aa_flag == 'unresolved' or aa_flag == 'mutation':\n",
    "                appender[idx]['id_a_aa'] = a\n",
    "                appender[idx]['id_a_pos'] = a_idx\n",
    "                appender[idx]['id_b_aa'] = b\n",
    "                appender[idx]['id_b_pos'] = b_idx\n",
    "                a_idx += 1\n",
    "                b_idx += 1\n",
    "\n",
    "            if aa_flag == 'deletion':\n",
    "                appender[idx]['id_a_aa'] = a\n",
    "                appender[idx]['id_a_pos'] = a_idx\n",
    "                a_idx += 1\n",
    "\n",
    "            if aa_flag == 'insertion':\n",
    "                appender[idx]['id_b_aa'] = b\n",
    "                appender[idx]['id_b_pos'] = b_idx\n",
    "                b_idx += 1\n",
    "            \n",
    "            idx += 1\n",
    "\n",
    "    alignment_df = pd.DataFrame.from_dict(appender, orient='index')\n",
    "    alignment_df = alignment_df[['id_a', 'id_b', 'type', 'id_a_aa', 'id_a_pos', 'id_b_aa', 'id_b_pos']].fillna(value=np.nan)\n",
    "    \n",
    "    return alignment_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id_a</th>\n",
       "      <th>id_b</th>\n",
       "      <th>type</th>\n",
       "      <th>id_a_aa</th>\n",
       "      <th>id_a_pos</th>\n",
       "      <th>id_b_aa</th>\n",
       "      <th>id_b_pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>deletion</td>\n",
       "      <td>M</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>deletion</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>deletion</td>\n",
       "      <td>Q</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>deletion</td>\n",
       "      <td>T</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>P</td>\n",
       "      <td>5</td>\n",
       "      <td>P</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>A</td>\n",
       "      <td>6</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>F</td>\n",
       "      <td>7</td>\n",
       "      <td>F</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>D</td>\n",
       "      <td>8</td>\n",
       "      <td>D</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>K</td>\n",
       "      <td>9</td>\n",
       "      <td>K</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>P</td>\n",
       "      <td>10</td>\n",
       "      <td>P</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>K</td>\n",
       "      <td>11</td>\n",
       "      <td>K</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>V</td>\n",
       "      <td>12</td>\n",
       "      <td>V</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>E</td>\n",
       "      <td>13</td>\n",
       "      <td>E</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>L</td>\n",
       "      <td>14</td>\n",
       "      <td>L</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>H</td>\n",
       "      <td>15</td>\n",
       "      <td>H</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>V</td>\n",
       "      <td>16</td>\n",
       "      <td>V</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>H</td>\n",
       "      <td>17</td>\n",
       "      <td>H</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>L</td>\n",
       "      <td>18</td>\n",
       "      <td>L</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>D</td>\n",
       "      <td>19</td>\n",
       "      <td>D</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>G</td>\n",
       "      <td>20</td>\n",
       "      <td>G</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>S</td>\n",
       "      <td>21</td>\n",
       "      <td>S</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>I</td>\n",
       "      <td>22</td>\n",
       "      <td>I</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>K</td>\n",
       "      <td>23</td>\n",
       "      <td>K</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>P</td>\n",
       "      <td>24</td>\n",
       "      <td>P</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>E</td>\n",
       "      <td>25</td>\n",
       "      <td>E</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>T</td>\n",
       "      <td>26</td>\n",
       "      <td>T</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>I</td>\n",
       "      <td>27</td>\n",
       "      <td>I</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>L</td>\n",
       "      <td>28</td>\n",
       "      <td>L</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>Y</td>\n",
       "      <td>29</td>\n",
       "      <td>Y</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>Y</td>\n",
       "      <td>30</td>\n",
       "      <td>Y</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>334</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>L</td>\n",
       "      <td>335</td>\n",
       "      <td>L</td>\n",
       "      <td>331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>335</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>P</td>\n",
       "      <td>336</td>\n",
       "      <td>P</td>\n",
       "      <td>332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>336</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>E</td>\n",
       "      <td>337</td>\n",
       "      <td>E</td>\n",
       "      <td>333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>337</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>D</td>\n",
       "      <td>338</td>\n",
       "      <td>D</td>\n",
       "      <td>334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>338</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>E</td>\n",
       "      <td>339</td>\n",
       "      <td>E</td>\n",
       "      <td>335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>K</td>\n",
       "      <td>340</td>\n",
       "      <td>K</td>\n",
       "      <td>336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>340</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>R</td>\n",
       "      <td>341</td>\n",
       "      <td>R</td>\n",
       "      <td>337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>341</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>E</td>\n",
       "      <td>342</td>\n",
       "      <td>E</td>\n",
       "      <td>338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>342</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>L</td>\n",
       "      <td>343</td>\n",
       "      <td>L</td>\n",
       "      <td>339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>343</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>L</td>\n",
       "      <td>344</td>\n",
       "      <td>L</td>\n",
       "      <td>340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>344</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>D</td>\n",
       "      <td>345</td>\n",
       "      <td>D</td>\n",
       "      <td>341</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>345</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>L</td>\n",
       "      <td>346</td>\n",
       "      <td>L</td>\n",
       "      <td>342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>346</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>L</td>\n",
       "      <td>347</td>\n",
       "      <td>L</td>\n",
       "      <td>343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>347</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>Y</td>\n",
       "      <td>348</td>\n",
       "      <td>Y</td>\n",
       "      <td>344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>348</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>K</td>\n",
       "      <td>349</td>\n",
       "      <td>K</td>\n",
       "      <td>345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>A</td>\n",
       "      <td>350</td>\n",
       "      <td>A</td>\n",
       "      <td>346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>350</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>Y</td>\n",
       "      <td>351</td>\n",
       "      <td>Y</td>\n",
       "      <td>347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>351</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>G</td>\n",
       "      <td>352</td>\n",
       "      <td>G</td>\n",
       "      <td>348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>352</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>M</td>\n",
       "      <td>353</td>\n",
       "      <td>M</td>\n",
       "      <td>349</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>353</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>P</td>\n",
       "      <td>354</td>\n",
       "      <td>P</td>\n",
       "      <td>350</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>354</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>P</td>\n",
       "      <td>355</td>\n",
       "      <td>P</td>\n",
       "      <td>351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>355</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>S</td>\n",
       "      <td>356</td>\n",
       "      <td>S</td>\n",
       "      <td>352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>A</td>\n",
       "      <td>357</td>\n",
       "      <td>A</td>\n",
       "      <td>353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>S</td>\n",
       "      <td>358</td>\n",
       "      <td>S</td>\n",
       "      <td>354</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>358</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>A</td>\n",
       "      <td>359</td>\n",
       "      <td>A</td>\n",
       "      <td>355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>359</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>G</td>\n",
       "      <td>360</td>\n",
       "      <td>G</td>\n",
       "      <td>356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>360</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>Q</td>\n",
       "      <td>361</td>\n",
       "      <td>Q</td>\n",
       "      <td>357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>361</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>N</td>\n",
       "      <td>362</td>\n",
       "      <td>N</td>\n",
       "      <td>358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>362</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>match</td>\n",
       "      <td>L</td>\n",
       "      <td>363</td>\n",
       "      <td>L</td>\n",
       "      <td>359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>363</th>\n",
       "      <td>P00813-1</td>\n",
       "      <td>3iar.A</td>\n",
       "      <td>insertion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>360</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>364 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         id_a    id_b       type id_a_aa  id_a_pos id_b_aa  id_b_pos\n",
       "0    P00813-1  3iar.A   deletion       M         1     NaN       NaN\n",
       "1    P00813-1  3iar.A   deletion       A         2     NaN       NaN\n",
       "2    P00813-1  3iar.A   deletion       Q         3     NaN       NaN\n",
       "3    P00813-1  3iar.A   deletion       T         4     NaN       NaN\n",
       "4    P00813-1  3iar.A      match       P         5       P         1\n",
       "5    P00813-1  3iar.A      match       A         6       A         2\n",
       "6    P00813-1  3iar.A      match       F         7       F         3\n",
       "7    P00813-1  3iar.A      match       D         8       D         4\n",
       "8    P00813-1  3iar.A      match       K         9       K         5\n",
       "9    P00813-1  3iar.A      match       P        10       P         6\n",
       "10   P00813-1  3iar.A      match       K        11       K         7\n",
       "11   P00813-1  3iar.A      match       V        12       V         8\n",
       "12   P00813-1  3iar.A      match       E        13       E         9\n",
       "13   P00813-1  3iar.A      match       L        14       L        10\n",
       "14   P00813-1  3iar.A      match       H        15       H        11\n",
       "15   P00813-1  3iar.A      match       V        16       V        12\n",
       "16   P00813-1  3iar.A      match       H        17       H        13\n",
       "17   P00813-1  3iar.A      match       L        18       L        14\n",
       "18   P00813-1  3iar.A      match       D        19       D        15\n",
       "19   P00813-1  3iar.A      match       G        20       G        16\n",
       "20   P00813-1  3iar.A      match       S        21       S        17\n",
       "21   P00813-1  3iar.A      match       I        22       I        18\n",
       "22   P00813-1  3iar.A      match       K        23       K        19\n",
       "23   P00813-1  3iar.A      match       P        24       P        20\n",
       "24   P00813-1  3iar.A      match       E        25       E        21\n",
       "25   P00813-1  3iar.A      match       T        26       T        22\n",
       "26   P00813-1  3iar.A      match       I        27       I        23\n",
       "27   P00813-1  3iar.A      match       L        28       L        24\n",
       "28   P00813-1  3iar.A      match       Y        29       Y        25\n",
       "29   P00813-1  3iar.A      match       Y        30       Y        26\n",
       "..        ...     ...        ...     ...       ...     ...       ...\n",
       "334  P00813-1  3iar.A      match       L       335       L       331\n",
       "335  P00813-1  3iar.A      match       P       336       P       332\n",
       "336  P00813-1  3iar.A      match       E       337       E       333\n",
       "337  P00813-1  3iar.A      match       D       338       D       334\n",
       "338  P00813-1  3iar.A      match       E       339       E       335\n",
       "339  P00813-1  3iar.A      match       K       340       K       336\n",
       "340  P00813-1  3iar.A      match       R       341       R       337\n",
       "341  P00813-1  3iar.A      match       E       342       E       338\n",
       "342  P00813-1  3iar.A      match       L       343       L       339\n",
       "343  P00813-1  3iar.A      match       L       344       L       340\n",
       "344  P00813-1  3iar.A      match       D       345       D       341\n",
       "345  P00813-1  3iar.A      match       L       346       L       342\n",
       "346  P00813-1  3iar.A      match       L       347       L       343\n",
       "347  P00813-1  3iar.A      match       Y       348       Y       344\n",
       "348  P00813-1  3iar.A      match       K       349       K       345\n",
       "349  P00813-1  3iar.A      match       A       350       A       346\n",
       "350  P00813-1  3iar.A      match       Y       351       Y       347\n",
       "351  P00813-1  3iar.A      match       G       352       G       348\n",
       "352  P00813-1  3iar.A      match       M       353       M       349\n",
       "353  P00813-1  3iar.A      match       P       354       P       350\n",
       "354  P00813-1  3iar.A      match       P       355       P       351\n",
       "355  P00813-1  3iar.A      match       S       356       S       352\n",
       "356  P00813-1  3iar.A      match       A       357       A       353\n",
       "357  P00813-1  3iar.A      match       S       358       S       354\n",
       "358  P00813-1  3iar.A      match       A       359       A       355\n",
       "359  P00813-1  3iar.A      match       G       360       G       356\n",
       "360  P00813-1  3iar.A      match       Q       361       Q       357\n",
       "361  P00813-1  3iar.A      match       N       362       N       358\n",
       "362  P00813-1  3iar.A      match       L       363       L       359\n",
       "363  P00813-1  3iar.A  insertion     NaN       NaN       A       360\n",
       "\n",
       "[364 rows x 7 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_alignment_allpos_df(alignment_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
